mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+}
+
+func (txmp *TxPool) evictTx(wtx *wrappedTx) {
+	txmp.store.remove(wtx.key)
+	txmp.metrics.EvictedTxs.Add(1)
+	txmp.evictedTxs.Push(wtx)
+	txmp.logger.Debug(
+		"evicted valid existing transaction; mempool full",
+		"old_tx", fmt.Sprintf("%X", wtx.key),
+		"old_priority", wtx.priority,
+	)
+}
+
+// handleRecheckResult handles the responses from ABCI CheckTx calls issued
+// during the recheck phase of a block Update.  It removes any transactions
+// invalidated by the application.
+//
+// This method is NOT executed for the initial CheckTx on a new transaction;
+// that case is handled by addNewTransaction instead.
+func (txmp *TxPool) handleRecheckResult(wtx *wrappedTx, checkTxRes *abci.ResponseCheckTx) {
+	txmp.metrics.RecheckTimes.Add(1)
+
+	// If a postcheck hook is defined, call it before checking the result.
+	err := txmp.postCheck(wtx.tx, checkTxRes)
+
+	if checkTxRes.Code == abci.CodeTypeOK && err == nil {
+		// Note that we do not update the transaction with any of the values returned in
+		// recheck tx
+		return // N.B. Size of mempool did not change
+	}
+
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.priority,
+		"tx", fmt.Sprintf("%X", wtx.key),
+		"err", err,
+		"code", checkTxRes.Code,
+	)
+	txmp.store.remove(wtx.key)
+	txmp.rejectedTxCache.Push(wtx.key)
+	txmp.metrics.FailedTxs.Add(1)
+	txmp.metrics.Size.Set(float64(txmp.Size()))
+}
+
+// recheckTransactions initiates re-CheckTx ABCI calls for all the transactions
+// currently in the mempool. It reports the number of recheck calls that were
+// successfully initiated.
+//
+// Precondition: The mempool is not empty.
+// The caller must hold txmp.mtx exclusively.
+func (txmp *TxPool) recheckTransactions() {
+	if txmp.Size() == 0 {
+		panic("mempool: cannot run recheck on an empty mempool")
+	}
+	txmp.logger.Debug(
+		"executing re-CheckTx for all remaining transactions",
+		"num_txs", txmp.Size(),
+		"height", txmp.Height(),
+	)
+
+	// Collect transactions currently in the mempool requiring recheck.
+	wtxs := txmp.store.getAllTxs()
+
+	// Issue CheckTx calls for each remaining transaction, and when all the
+	// rechecks are complete signal watchers that transactions may be available.
+	go func() {
+		g, start := taskgroup.New(nil).Limit(2 * runtime.NumCPU())
+
+		for _, wtx := range wtxs {
+			wtx := wtx
+			start(func() error {
+				// The response for this CheckTx is handled by the default recheckTxCallback.
+				rsp, err := txmp.proxyAppConn.CheckTxSync(abci.RequestCheckTx{
+					Tx:   wtx.tx,
+					Type: abci.CheckTxType_Recheck,
+				})
+				if err != nil {
+					txmp.logger.Error("failed to execute CheckTx during recheck",
+						"err", err, "key", fmt.Sprintf("%x", wtx.key))
+				} else {
+					txmp.handleRecheckResult(wtx, rsp)
+				}
+				return nil
+			})
+		}
+		_ = txmp.proxyAppConn.FlushAsync()
+
+		// When recheck is complete, trigger a notification for more transactions.
+		_ = g.Wait()
```

Comment:
question: how are the taskgroups handle with shutdowns? Are they blocking? if not, are there any concerns with unclean shutdowns with applying txns?

Reply:
Yeah having contexts would be nice. In later tendermint version we added contexts to the ABCI calls. Currently this is blocking.